io/common/include/detray/io/geometry_reader.hpp

```diff
@@ -0,0 +1,33 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s)
+#include "detray/io/payloads.hpp"
+
+// System include(s)
+#include <string>
+#include <vector>
+
+namespace detray {
+
+/// @brief Abstract base class for tracking geometry readers
+class geometry_reader {
+
+    public:
+    /// Default constructor
+    geometry_reader() = default;
+
+    /// Default destructor
+    virtual ~geometry_reader() {}
+
+    /// Reads the data from a given file into a collection of volume payloads
+    virtual const std::vector<volume_payload>& read(const std::string&) = 0;
```

Comment:
Tbh, the reading part is still in the brainstorming phase and this file does not actually belong to this PR. Also the comments are a bit misleading/incomplete perhaps. The string that is passed in this interface is not the full filename, but the global detector name. The plan is, to be able to plug readers/writers for detector components into a global detector reader/writer. The io components then decide for themselves what the file name is going to be and how to open the file, depending on what part of the detector data they are handling and on the file format. That's why I put the file handling biolerplate into this function instead of putting it into the final detector writer, which would otherwise have to handle files/streams for an a priori unknown number and type of registered readers/writers.  Maybe this can be handled more elegantly, idk. In the end, I planned to have smth conceptually like this (even though the pattern for that is not part of this PR):

```
    detray::detector_writer<detector_t> det_writer;

    det_writer.template add<json_geometry_writer>(); // Could also be a template param, since geo io is mandatory
    // det_writer.template add<json_simple_material_writer>();
    // det_writer.template add<csv_simple_material_writer>();
    det_writer.template add<json_material_map_writer>();
    det_writer.template add<json_grid_writer>();
    det_writer.template add<json_kd_tree_writer>();
    ....

   det_writer.write(det, "toy_geometry");
````

Depending on what components a detector actually has (i.e. what components are found in the files), it should be able to dynamically register the correct readers/writers and, later, the corresponding builders. What I really want to avoid, is a monolithic io class that is locked into a specific way of writing files and can only write detectors of a certain topology.